src/program/calibration.rs

```diff
@@ -91,15 +120,90 @@ impl CalibrationSet {
                                     }
                                 }
                             }
+
+                            instruction.apply_to_expressions(|expr| {
+                                let previous = std::mem::replace(expr, Expression::PiConstant);
+                                *expr = previous.substitute_variables(&variable_expansions);
+                            })
                         }
 
                         Some(instructions)
                     }
                     None => None,
                 }
             }
+            Instruction::Measurement(Measurement { qubit, target }) => {
+                let mut matching_calibrations_with_precedence: Vec<(
+                    u32,
+                    &MeasureCalibrationDefinition,
+                )> = self
+                    .measure_calibrations
+                    .iter()
+                    .filter_map(|cal| match &cal.qubit {
+                        Some(cal_qubit) if qubit == cal_qubit => Some((1, cal)),
+                        None => Some((0, cal)),
+                        _ => None,
+                    })
+                    .collect();
+
+                // The matching calibration is the last-specified one that matched the target qubit (if any),
+                // or otherwise the last-specified one that specified no qubit.
+                matching_calibrations_with_precedence.sort_by_key(|(precedence, _)| *precedence);
+                let matching_calibration = matching_calibrations_with_precedence.last();
```

Comment:
```suggestion
                let mut matching_calibration = None;
                let mut found_qubit_match = false;
                for cal in self.measure_calibrations {
                    if let Some(cal_qubit) = &cal.qubit {
                        if cal_qubit == qubit {
                            found_qubit_match = true;
                            matching_calibration = Some(cal);
                        }
                    } else if !found_qubit_match {
                        matching_calibration = Some(cal);
                    }
                }
```

This is more obvious to my eyes, but that's certainly subjective. Another approach is maybe something like this which is more functional but less performant:


```suggestion
                let mut matching_calibration = self.measure_calibrations.iter().rev().find(|cal| {
                    if let Some(cal_qubit) = &cal.qubit {
                        cal_qubit == qubit
                    } else {
                        false
                    }
                });
                if matching_calibration.is_none() {
                    matching_calibration = self
                        .measure_calibrations
                        .iter()
                        .rev()
                        .find(|cal| cal.qubit.is_none());
                }
```

Reply:
Absolutely right, thank you for keeping me honest here and for the suggestions! I applied your first suggestion with some tweaks for ordering (the latest calibration is the one to apply of those with equal precedence).